### PR TITLE
[flutter_plugin_tools] Improve package targeting

### PR DESCRIPTION
**Pull request by stuartmorgan**
Improve package targeting:
- `--run-on-changed-packages` now includes only changed packages in a federated plugin, not all packages. Include all packages isn't useful since (without changes that would cause them to be included anyway) package dependencies are not path-based between packages.
- `--packages` now allows specifying package names of federated plugins. E.g., `--packages=path_provider_ios` will now work, instead of requiring `--packages=path_provider/path_provider_ios`. The fully qualified form still works as well (and is still needed for app-facing packages to disambiguate from the plugin group).

Fixes https://github.com/flutter/flutter/issues/94618

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: script/tool/CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,11 @@
 ## NEXT
 
 - Ensures that `firebase-test-lab` runs include an `integration_test` runner.
+- `--packages` now allows using a federated plugin's package as a target without
+  fully specifying it (if it is not the same as the plugin's name). E.g.,
```
Comment (gaaclarke) [inline script/tool/CHANGELOG.md:5]: Should this information be in the help page for the tool?

File: script/tool/lib/src/common/plugin_command.dart [hunk 2]
```diff
@@ -432,17 +434,48 @@ abstract class PluginCommand extends Command<void> {
...
-  Set<String> _getChangedPackages(List<String> changedFiles) {
+  Set<String> _getChangedPackageNames(List<String> changedFiles) {
     final Set<String> packages = <String>{};
+
+    // A helper function that returns true if candidateName looks like an
+    // implementation package of a plugin called pluginName. Used to determine
+    // if .../packages/parentName/candidatePackageName/...
+    // looks like a path in a federated plugin package (candidatePackageName)
+    // rather than a top-level package (parentName).
+    bool isFederatedPackage(String candidatePackageName, String parentName) {
```
Comment (gaaclarke) [inline script/tool/lib/src/common/plugin_command.dart:453]: s/isFederatedPackage/_isFederatedPackage

File: script/tool/lib/src/common/plugin_command.dart [hunk 3]
```diff
@@ -432,17 +434,48 @@ abstract class PluginCommand extends Command<void> {
...
   // The paths must use POSIX separators (e.g., as provided by git output).
-  Set<String> _getChangedPackages(List<String> changedFiles) {
+  Set<String> _getChangedPackageNames(List<String> changedFiles) {
     final Set<String> packages = <String>{};
+
+    // A helper function that returns true if candidateName looks like an
+    // implementation package of a plugin called pluginName. Used to determine
+    // if .../packages/parentName/candidatePackageName/...
+    // looks like a path in a federated plugin package (candidatePackageName)
+    // rather than a top-level package (parentName).
```
Comment (gaaclarke) [inline script/tool/lib/src/common/plugin_command.dart:452]: You might as well make this a docstring so you get compiler support about variable names and what not.  There is already a mixmatch between "candidateName" and "candidatePackageName"

**Review by gaaclarke — APPROVED** [review]
I have a few nits on documentation, code and tests look good.

File: script/tool/CHANGELOG.md (hunk 1 above)
Comment (stuartmorgan) [reply script/tool/CHANGELOG.md:5]: Good idea, done.

File: script/tool/lib/src/common/plugin_command.dart (hunk 3 above)
Comment (stuartmorgan) [reply script/tool/lib/src/common/plugin_command.dart:452]: This is a locally scoped function, which doesn't seem to support docstrings (at least not in VS Code). I fixed `candidateName` though.

File: script/tool/lib/src/common/plugin_command.dart (hunk 2 above)
Comment (stuartmorgan) [reply script/tool/lib/src/common/plugin_command.dart:453]: Locals shouldn't be named with underscores.

**Comment by stuartmorgan** [comment]
Overriding incorrect submit-queue check; it's pulling status from the previous publish run, not the current (successful) run.

**Pull request merged**